### PR TITLE
Use smart pointers in GUIFormSpecMenu

### PR DESCRIPTION
**Pull request by Desour**
- Goal of the PR: Improve code
- How does the PR work?
Instead of storing raw pointers to gui elements and grabbing dropping them manually, `irr_ptr` (not used yet, but added some time ago by @numberZero for this purpose) are used.
`m_selected_item` is a `unique_ptr` now.
I've also fixed some style here and there, removed `m_checkboxes` and `m_scrollbars` which were unused and replaced a dynamic allocation of an SEvent.

## To do

This PR is a Ready for Review.
Note that there's still a todo comment to be fixed, see https://github.com/minetest/minetest/pull/9487#discussion_r389392014. Please help making the decision there / decide for me.

## How to test

Look for memory leaks and co.
There shouldn't be any `drop()`s anymore.
Note that the constructor for `irr_ptr` (and also `reset`) does not grab. Hence we have to grab if we get the element by some `add...` function. If a raw pointer to an element is stored somewhere (shouldn't happen anywhere) (by store I mean that it's later dropped again there), we also have to grab.

### CODE REVIEW

**Comment by rubenwardy** [comment]
I'm so glad to see this! Thanks

**Comment by appgurueu** [comment]
Travis is complaining about [GUIFormSpecMenu:1021](https://travis-ci.org/minetest/minetest/jobs/659691134#L3120) and wants you to "use emplace_back instead of push_back"

**Comment by Desour** [comment]
> Travis is complaining about [GUIFormSpecMenu:1021](https://travis-ci.org/minetest/minetest/jobs/659691134#L3120) and wants you to "use emplace_back instead of push_back"

- Since when does travis complain so much everywhere?!
- Why would I want to use `emplace_back`? Automatically copying should be better than manually getting the raw pointer and emplacing a new smart pointer. (A `std::move` could be used there btw., should I do that (one fewer grab and drop)?)

**Comment by rubenwardy** [comment]
> Why would I want to use emplace_back? Automatically copying should be better than manually getting the raw pointer and emplacing a new smart pointer

Remember the copy constructor. You can emplace a smart pointer into a vector if it implements the copy constructor

> A std::move could be used there btw., should I do that (one fewer grab and drop)?

Yes, you should prefer std::move where it makes sense

**Comment by Desour** [comment]
> Remember the copy constructor. You can emplace a smart pointer into a vector if it implements the copy constructor

Afaik, `emplace` doesn't use the copy constructor, it constructs in-place. `push_back` uses the copy constructor or move constructor. See also https://en.cppreference.com/w/cpp/container/vector/push_back and https://en.cppreference.com/w/cpp/container/vector/emplace_back.

Hm, `std::move` could also be used everywhere where a `FieldSpec` is pushed back to `m_fields`.

File: src/gui/guiFormSpecMenu.cpp [hunk 1]
```diff
@@ -2799,19 +2765,18 @@ void GUIFormSpecMenu::regenerateGui(v2u32 screensize)
 
 	// Add tooltip
 	{
-		assert(!m_tooltip_element);
+		assert(!m_tooltip_element); //todo: do we still need this assert? (sadly irr_ptr does not allow to set a deleter and m_tooltip_element has to be ->remove()d)
```
Comment (Desour) [inline src/gui/guiFormSpecMenu.cpp]: When `m_tooltip_element` is dropped (see `removeChildren`), it also has to be removed from its parent. Hence I'm unsure if this assert can be safely removed.

**Comment by Desour** [comment]
Rebased.

File: src/gui/guiFormSpecMenu.cpp [hunk 2]
```diff
@@ -3423,7 +3380,7 @@ void GUIFormSpecMenu::updateSelectedItem()
 
 	// If craftresult is nonempty and nothing else is selected, select it now.
 	if (!m_selected_item) {
-		for (const GUIInventoryList *e : m_inventorylists) {
+		for (const irr_ptr<GUIInventoryList> &e : m_inventorylists) {
```
Comment (Desour) [inline src/gui/guiFormSpecMenu.cpp:3383]: Not that there's a change here.
`const irr_ptr<T> e` is like `T *const e`, `irr_ptr<const T> e` would be like `const T *e`.
I could add some `const`s everywhere, if requested. (Ew, mutability by default.)

**Comment by raymoo** [comment]
>Afaik, emplace doesn't use the copy constructor, it constructs in-place.

`push_back` constructs in-place too, using the copy constructor (though some specific usage of `push_back` may not be constructing in place, if an object is constructed to be passed into `push_back`). The difference with `emplace_back` is that it will choose the constructor to use based on the arguments, to allow use of other constructors. If you pass it the arguments for the copy constructor, it will use the copy constructor like `push_back` would have.

**Comment by Desour** [comment]
Rebased.

**Comment by Desour** [comment]
I've decided to keep the one assert there.
PR is ready.

**Comment by Desour** [comment]
Squashed and rebased.

File: src/gui/guiFormSpecMenu.cpp [hunk 3]
```diff
@@ -907,8 +886,8 @@ void GUIFormSpecMenu::parseAnimatedImage(parserData *data, const std::string &el
 
 	core::rect<s32> rect = core::rect<s32>(pos, pos + geom);
 
-	GUIAnimatedImage *e = new GUIAnimatedImage(Environment, this, spec.fid,
-		rect, texture_name, frame_count, frame_duration, m_tsrc);
+	irr_ptr<GUIAnimatedImage> e(new GUIAnimatedImage(Environment, data->current_parent,
```
Comment (SmallJoker) [inline src/gui/guiFormSpecMenu.cpp:889]: Changed behaviour - is this a bugfix?

File: src/gui/guiFormSpecMenu.cpp [hunk 4]
```diff
@@ -4414,7 +4369,6 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
 		// If m_selected_amount has been decreased to zero, deselect
 		if (m_selected_amount == 0) {
 			m_selected_swap.clear();
-			delete m_selected_item;
 			m_selected_item = nullptr;
```
Comment (SmallJoker) [inline src/gui/guiFormSpecMenu.cpp:4372]: How does this work? `ItemSpec` does not inherit from `ReferenceCounted`, so would it ever be deleted?

Comment (SmallJoker) [inline src/gui/guiFormSpecMenu.cpp:4372]: Because this is not a plain pointer please use `m_selected_item.reset();` here.

**Comment by SmallJoker** [comment]
Works. No new leaks detected using valgrind.

File: src/gui/guiFormSpecMenu.cpp (hunk 3 above)
Comment (Desour) [reply src/gui/guiFormSpecMenu.cpp:889]: Yes. Changing the parent for this element was probably forgotten in a rebase.

File: src/gui/guiFormSpecMenu.cpp (hunk 4 above)
Comment (Desour) [reply src/gui/guiFormSpecMenu.cpp]: Done.

**Comment by Desour** [comment]
Rebased.

**Comment by Desour** [comment]
Uh, looks like you shouldn't rename branches of PRs in github.
I wanted to rebase and renamed the old branch to `fs_irr_ptr_before_rebase`. The rebased version would be available under `fs_irr_ptr`.
Anyway, I am frightened of the silent mistakes the rebase might've caused. So this is closed for now, I can reopen when there's more interest.

**Pull request closed without merging**